// @skyway-sdk/sfu-bot/dist/forwarding.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Forwarding = void 0;
const common_1 = require("@skyway-sdk/common");
class Forwarding {
    /**@internal */
    constructor(configure, originPublication, relayingPublication) {
        this.configure = configure;
        this.originPublication = originPublication;
        this.relayingPublication = relayingPublication;
        this.state = 'started';
        /** @description [japanese] forwardingが終了された時に発火するイベント */
        this.onStopped = new common_1.Event();
    }
    get id() {
        return this.relayingPublication.id;
    }
    /**@private */
    _stop() {
        this.state = 'stopped';
        this.onStopped.emit();
    }
    /**@internal */
    toJSON() {
        return {
            id: this.id,
            configure: this.configure,
            originPublication: this.originPublication,
            relayingPublication: this.relayingPublication,
        };
    }
}
exports.Forwarding = Forwarding;
